### PR TITLE
rgw: lua, add support for lua packages to the lua manager

### PR DESCRIPTION
**Pull request by kalebskeithley**
Refactor lua manager, adding new {add,remove,install}_package(), and
list_packages() methods; and changing get(), put(), del() methods
to {get,put,del}_script().  Renaming class from *LuaScriptManager
to *LuaManager to be consistent with dual *script and *package
functionality.

Note: I considered keeping the existing *LuaScriptManager classes
as is and adding *LuaPackageManager classes, but that seemed/seems
like overkill to me.

Signed-off-by: Kaleb S. KEITHLEY <kkeithle@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_sal_rados.h [hunk 1]
```diff
@@ -868,17 +868,28 @@ class RadosMultipartWriter : public StoreWriter {
...
+  virtual int get_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) override;
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) override;
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) override;
+#ifdef WITH_RADOSGW_LUA_PACKAGES
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name, bool allow_compilation) override;
+  virtual int remove_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) override;
+  virtual int list_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& packages) override;
+  virtual int install_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& failed_packages, std::string& output) override;
+#else
+    virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name, bool allow_compilation) { return -ENOENT; }
```
Comment (dang) [inline src/rgw/rgw_sal_rados.h:888]: Suggestion: Put these in StoreLuaManager, then you don't need the #else case here, and you can leave the stub implementations out of DBStore and MotrStore.

**Comment by cbodley** [comment]
looks pretty good! i think the stuff involving 'boost/process.hpp' that shells out to `luarocks` should probably stay above the zipper line, so we don't duplicate that part in each backend. only the parts that rely on `librados` need to move below

**Comment by kalebskeithley** [comment]
jenkins test make check

**Comment by kalebskeithley** [comment]
jenkins test api

**Comment by kalebskeithley** [comment]
jenkins test api

File: src/rgw/rgw_lua.cc [hunk 2]
```diff
@@ -70,188 +65,53 @@ std::string script_oid(context ctx, const std::string& tenant) {
...
-    } while (c.running() && std::getline(is, line));
+  return lua_mgr->list_packages(dpp, y, packages);
+}
 
-    c.wait();
-    if (c.exit_code()) {
-      failed_packages.insert(package);
-    }
-  }
+int install_packages(const DoutPrefixProvider *dpp, rgw::sal::Store* store, optional_yield y, packages_t& failed_packages, std::string& output)
```
Comment (yuvalif) [inline src/rgw/rgw_lua.cc:109]: this does not have any RADOS dependency.
it just calls into `list_packages()`

File: src/rgw/rgw_lua.h [hunk 3]
```diff
@@ -36,23 +38,23 @@ int read_script(const DoutPrefixProvider *dpp, rgw::sal::Store* store, const std
 // delete the stored lua script from a context
 int delete_script(const DoutPrefixProvider *dpp, rgw::sal::Store* store, const std::string& tenant, optional_yield y, context ctx);
 
-#ifdef WITH_RADOSGW_LUA_PACKAGES
 #include <set>
 
 using packages_t = std::set<std::string>;
 
+#ifdef WITH_RADOSGW_LUA_PACKAGES
```
Comment (yuvalif) [inline src/rgw/rgw_lua.h:45]: why not include lines 41 and 43 inside the `ifdef`?

File: src/rgw/rgw_main.cc [hunk 4]
```diff
@@ -483,11 +481,10 @@ int radosgw_Main(int argc, const char **argv)
     store->set_luarocks_path(luarocks_path+"/"+g_conf()->name.to_str());
   }
 #ifdef WITH_RADOSGW_LUA_PACKAGES
-  rgw::sal::RadosStore *rados = dynamic_cast<rgw::sal::RadosStore*>(store);
-  if (rados) { /* Supported for only RadosStore */
+  if (store->get_name() == "rados") { /* Supported for only RadosStore */
```
Comment (yuvalif) [inline src/rgw/rgw_main.cc:484]: if we move the implementation of `install_packages` to the base class, we can remove the RADOS check:
```cpp
if (store->get_name() == "rados")
```

File: src/rgw/rgw_sal.h [hunk 5]
```diff
@@ -1498,20 +1499,29 @@ class Zone {
...
   /** Get a script named with the given key from the backing store */
-  virtual int get(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) = 0;
+  virtual int get_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) = 0;
   /** Put a script named with the given key to the backing store */
-  virtual int put(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) = 0;
+  virtual int put_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, const std::string& script) = 0;
   /** Delete a script named with the given key from the backing store */
-  virtual int del(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  virtual int del_script(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key) = 0;
+  /** Add a lua package */
```
Comment (yuvalif) [inline src/rgw/rgw_sal.h:1517]: why not protect these under the `ifdef`?

File: src/rgw/rgw_sal.h [hunk 6]
```diff
@@ -1498,20 +1499,29 @@ class Zone {
...
+  virtual int add_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name, bool allow_compilation) { return -ENOENT; }
```
Comment (yuvalif) [inline src/rgw/rgw_sal.h:1518]: why do we have default implementation here and not make them pure virtual as with the scripts?

File: src/rgw/rgw_sal.h [hunk 7]
```diff
@@ -1498,20 +1499,29 @@ class Zone {
...
+  /** Remove a lua package */
+  virtual int remove_package(const DoutPrefixProvider* dpp, optional_yield y, const std::string& package_name) { return -ENOENT; }
+  /** List lua packages */
+  virtual int list_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& packages) { return -ENOENT; }
+  /** Install lua packages */
+  virtual int install_packages(const DoutPrefixProvider* dpp, optional_yield y, rgw::lua::packages_t& failed_packages, std::string& output) { return -ENOENT; }
```
Comment (yuvalif) [inline src/rgw/rgw_sal.h:1524]: as mentioned here: https://github.com/ceph/ceph/pull/46980/files#r917355672

we can have the default implementation

File: src/rgw/rgw_sal_filter.h [hunk 8]
```diff
@@ -610,9 +610,9 @@ class FilterZone : public Zone {
     virtual const std::string& get_realm_id() = 0;
 };
 
-class FilterLuaScriptManager : public LuaScriptManager {
+class FilterLuaManager : public LuaManager {
 public:
-  virtual ~FilterLuaScriptManager() = default;
+  virtual ~FilterLuaManager() = default;
 
   virtual int get(const DoutPrefixProvider* dpp, optional_yield y, const std::string& key, std::string& script) = 0;
```
Comment (yuvalif) [inline src/rgw/rgw_sal_filter.h:617]: shouldn't these be:
`get_script()`, `put_script()` and `del_script()` ?

Comment (dang) [reply src/rgw/rgw_sal_filter.h:617]: Yeah.  Filter's not being built yet, so it won't get caught.

File: src/rgw/rgw_sal.h (hunk 7 above)
Comment (dang) [reply src/rgw/rgw_sal.h:1524]: but in rgw_sal_store.h

File: src/rgw/rgw_sal.h (hunk 5 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal.h:1517]: it keeps the interface the same regardless of how it's built.

I thought that was conceptually cleaner. Theoretically less likely to cause strange crashes if somehow you get a mismatched rgw and plugin compiled different ways, i.e.  radosgw compiled *with* the ifdef, but plugin compiled *without* the ifdef

File: src/rgw/rgw_lua.cc (hunk 2 above)
Comment (kalebskeithley) [reply src/rgw/rgw_lua.cc:109]: I don't understand what you're trying to tell me here.

File: src/rgw/rgw_main.cc (hunk 4 above)
Comment (kalebskeithley) [reply src/rgw/rgw_main.cc:484]: Just so I'm clear, you want me to move the implementation (currently in RadosLuaManager in rgw_sal_rados.cc) to StoreLuaManager in rgw_sal_store.h (or in a new rgw_sal_store.cc) instead?

File: src/rgw/rgw_sal_filter.h (hunk 8 above)
Comment (kalebskeithley) [reply src/rgw/rgw_sal_filter.h]: I thought I marked this as resolved.

File: src/rgw/rgw_main.cc (hunk 4 above)
Comment (kalebskeithley) [reply src/rgw/rgw_main.cc:484]: rgw_sal.h  declares the pure virtual interface.
rgw_sal_store.{h,cc} defines default implementations that return -ENOENT.  (Or so I thought.)
rgw_sal_rados.cc defines the RADOS implementations.
rgw_sal_{dbstore,motr,...} have no implemenations at this time.

Comment (yuvalif) [reply src/rgw/rgw_main.cc:484]: ok. thanks.
so, the default implementation should go in the "sal_store"

File: src/rgw/rgw_sal.h (hunk 7 above)
Comment (dang) [reply src/rgw/rgw_sal.h]: "sal" is the base interface for anything.  "sal_store" is the base implementation for stores (actual backing stores, living at the bottom of a stack). "sal_filter" is the base implementation (still under development) for filters, which live somewhere in the middle of the stack.

File: src/rgw/rgw_sal.h (hunk 5 above)
Comment (dang) [reply src/rgw/rgw_sal.h:1517]: It should be okay.  Anything overridden will call the overriding functions, anything not overridden will use the base implementation.

Comment (yuvalif) [reply src/rgw/rgw_sal.h:1517]: but this will work only if there is a base implementation.
if we keep these functions as pure virtuals that won't compile

**Comment by yuvalif** [comment]
regarding: https://github.com/ceph/ceph/pull/46980#discussion_r921350107
if we want to keep the `ifdef` out of the base "sal" file, we would need a stub implementation in the "sal_store" file on the `else` part of the `ifdef`

**Comment by cbodley** [comment]
> regarding: [#46980 (comment)](https://github.com/ceph/ceph/pull/46980#discussion_r921350107)
> if we want to keep the `ifdef` out of the base "sal" file, we would need a stub implementation in the "sal_store" file on the `else` part of the `ifdef`

can you clarify? why do we need anything in sal_store?

**Comment by yuvalif** [comment]
> > regarding: [#46980 (comment)](https://github.com/ceph/ceph/pull/46980#discussion_r921350107)
> > if we want to keep the `ifdef` out of the base "sal" file, we would need a stub implementation in the "sal_store" file on the `else` part of the `ifdef`
> 
> can you clarify? why do we need anything in sal_store?

if we have the lua packages functions not under ifdef in "rgw_sal.h" and we set the flag to "false" we will have pure virtuals without implementation in the derived classes (assuming the implementation is under the ifdef in the derived classes).
so, what we can do is add default stubs to "rgw_store.h" (not under ifdef) so that the derived classes does compile

**Comment by cbodley** [comment]
the goal would be to remove the #ifdefs from sal entirely, not just the base class. backends can support storing a list of package names, even if the distro we're building for doesn't have access to `luarocks`

**Comment by kalebskeithley** [comment]
superceded by https://github.com/ceph/ceph/pull/47333

**Pull request closed without merging**